# Trabajo Final

Alumno: **Alberto García Parrado**

Github: https://github.com/albertogparrado/Algoritmos-VIU

# Problema 3. Combinar cifras y operaciones

- El problema consiste en analizar el siguiente problema y diseñar un algoritmo que lo resuelva.
- Disponemos de las 9 cifras del 1 al 9 (excluimos el cero) y de los 4 signos básicos de las
operaciones fundamentales: suma(+), resta(-), multiplicación(*) y división(/)
- Debemos combinarlos alternativamente sin repetir ninguno de ellos para obtener una
cantidad dada. Un ejemplo sería para obtener el **4:4+2-6/3*1 = 4**
- Debe analizarse el problema para encontrar todos los valores enteros posibles planteando las
siguientes cuestiones:
    - ¿Qué valor máximo y mínimo se pueden obtener según las condiciones del problema?
    - ¿Es posible encontrar todos los valores enteros posibles entre dicho mínimo y máximo ?
- Nota: Es posible usar la función de python “eval” para evaluar una expresión:

## Preguntas:

## **1. ¿Cuantas posibilidades hay sin tener en cuenta las restricciones?**

Posibilidades de las cifras (1,2,3,4,5,6,7,8,9):

Si permitimos repetir cifras, cada una de las 5 posiciones puede tomar cualquiera de los 9 valores disponibles.  
Esto se calcula como:

$$
9^5 = 59.049 \text{ combinaciones posibles de cifras}
$$

Posibilidades de los operadores (+, -, *, /):
Si también permitimos repetir operadores, y hay 4 posiciones de operador entre los 5 números, cada una puede tomar cualquiera de los 4 símbolos:

$$
4^4 = 256 \text{ combinaciones posibles de operadores}
$$

Multiplicando ambas cantidades:

$$
9^5 \times 4^4 = 59.049 \times 256 = 15.108.096 \text{ combinaciones totales}
$$

## **2.  ¿Cual es la estructura de datos que mejor se adapta al problema? Argumentalo.(Es posible que hayas elegido una al principio y veas la necesidad de cambiar, arguentalo).**

La estructuea que mejor encajaría para este problema sería la utilización de **listas** para la secuencia, el uso de **booleanos** para manejar si el dato se ha usado o no y variables tipo int para guardar los resultados.

## **3. Según el modelo para el espacio de soluciones ¿Cual es la función objetivo? ¿Es un problema de maximización o minimización?**

En este problema, el **espacio de soluciones** está formado por todas las posibles combinaciones de 5 cifras (con posibles repeticiones) y 4 operadores aritméticos, que generan expresiones de la forma:

$$
E = n_1 \; op_1 \; n_2 \; op_2 \; n_3 \; op_3 \; n_4 \; op_4 \; n_5
$$

La **función objetivo** es precisamente el **valor entero resultante de evaluar cada expresión generada**, ya que ese es el elemento que deseamos analizar y sobre el que queremos optimizar.

Dado que el enunciado solicita **calcular tanto el valor mínimo como el máximo que se pueden obtener en este espacio de soluciones**, se trata de un **problema de optimización combinatoria multiobjetivo**, en el que se busca el valor máximo posible (problema de **maximización**) como el valor mínimo (problema de **minimización**).

Ambos objetivos son relevantes y complementarios en este análisis, permitiendo estudiar el **rango de valores alcanzables** dado el conjunto de cifras y operaciones. Siendo un algoritmo clasificado como un **problema exacto (no heurístico)** que pretende encontrar el verdadero mínimo y máximo y no una aproximación.

## **4. Diseña un algoritmo para resolver el problema por fuerza bruta.**

In [90]:
import time
from itertools import permutations

def es_entero(x):
    return abs(x - int(x)) < 1e-9

def crear_combinaciones_de_operaciones(ops):
    return list(permutations(ops, 4))

def crear_combinaciones_numericas(numeros):
    return list(permutations(numeros, 5))

def crear_combinaciones_entre_numeros_y_operaciones(combinacion_operaciones, combinacion_numeros):
    combinaciones = []
    resultados_enteros = []

    total = len(combinacion_numeros) * len(combinacion_operaciones)
    print(f"🧮 Total de combinaciones entre números y operaciones: {total}")

    for numeros in combinacion_numeros:
        for operaciones in combinacion_operaciones:
            expresion = f"{numeros[0]} {operaciones[0]} {numeros[1]} {operaciones[1]} {numeros[2]} {operaciones[2]} {numeros[3]} {operaciones[3]} {numeros[4]}"
            try:
                calculo = eval(expresion)
            except ZeroDivisionError:
                continue

            if es_entero(calculo):
                resultados_enteros.append(int(calculo))
                combinaciones.append((numeros, operaciones, int(calculo)))

    if resultados_enteros:
        print(f"📉 Mínimo: {min(resultados_enteros)}")
        print(f"📈 Máximo: {max(resultados_enteros)}")
    else:
        print("⚠️ No se encontraron resultados enteros válidos.")

def resolver_fuerza_bruta(operaciones=['+', '-', '*', '/'], numeros=list(range(1, 10))):
    inicio = time.time()
    combinacion_operaciones = crear_combinaciones_de_operaciones(operaciones)
    print(f"ℹ️ Total de combinaciones de operaciones: {len(combinacion_operaciones)}")
    combinacion_numeros = crear_combinaciones_numericas(numeros)
    print(f"ℹ️ Total de combinaciones numéricas: {len(combinacion_numeros)}")
    crear_combinaciones_entre_numeros_y_operaciones(combinacion_operaciones, combinacion_numeros)
    print(f"⏱️ Tiempo fuerza bruta: {round(time.time() - inicio, 2)} segundos")

resolver_fuerza_bruta()

ℹ️ Total de combinaciones de operaciones: 24
ℹ️ Total de combinaciones numéricas: 15120
🧮 Total de combinaciones entre números y operaciones: 362880
📉 Mínimo: -69
📈 Máximo: 77
⏱️ Tiempo fuerza bruta: 1.23 segundos


## **5. Calcula la complejidad del algoritmo por fuerza bruta.**

El algoritmo por fuerza bruta evalúa todas las combinaciones de cifras y operaciones:

- 5 posiciones de cifras con 9 posibles valores cada una: 9^5.
- 4 posiciones de operaciones con 4 posibles valores cada una: 4^4 .

Por lo tanto, la complejidad total del algoritmo por fuerza bruta es:

$$
\boxed{O(9^5 \times 4^4) = O(n^5)}
$$

## **6. Diseña un algoritmo que mejore la complejidad del algortimo por fuerza bruta. Argumenta porque crees que mejora el algoritmo por fuerza bruta.**

Aunque el algoritmo por fuerza bruta es rápido en conjuntos pequeños, se vuelve ineficiente en escalas mayores por el elevado número de combinaciones posibles. Para mejorar este comportamiento he diseñado un **algoritmo Branch & Bound con poda (Poda de ramas)** que **reduce drásticamente el número de combinaciones a calcular**.

En el algoritmo inicial se usaba `eval(expresion)` para calcular toda la expresión, pero al separarlo paso a paso podemos tomar decisiones de **poda temprana** en función del valor parcial de la operación, descartando aquellas soluciones que sabemos que no aportarán mejoras.

**Mejoras implementadas:**

- Separación de operaciones paso a paso: se calcula cada operación individualmente, permitiendo tomar decisiones antes de terminar la expresión.
- Poda por divisiones inválidas: si aparece una división no entera o por cero, la rama se descarta inmediatamente.
- Si en pasos intermedios vemos que el valor parcial no puede superar el máximo actual (maximización) o no puede ser menor que el mínimo actual (minimización), podamos la rama y evitamos continuar.
- Se usa una clave normalizada para almacenar resultados de operaciones conmutativas (por ejemplo, `3 + 2 = 2 + 3`), evitando recalcular las mismas operaciones múltiples veces y reduciendo el número de operaciones simples necesarias. (Esta mejora sería para mejorar la performance y no la cantidad de operaciones)
- Se descartan combinaciones que multipliquen o dividan por 0 o 1 de forma innecesaria.
- Se descartan divisiones no enteras que no aportan soluciones.

De esta forma, el algoritmo mantiene su **exactitud** mientras **reduce de forma significativa el número de combinaciones a calcular**.

**Operaciones a mejorar: 362880**

In [91]:
# Poda

def es_entero(x):
    return abs(x - int(x)) < 1e-9

def clave_normalizada(a, op, b):
    return (min(a, b), op, max(a, b)) if op in ('+', '*') else (a, op, b)

def se_descarta_por_poda(operaciones, numeros, max_val, min_val):
    """
    Si en los primeros pasos ya vemos que el valor parcial
    no puede superar el máximo ni mejorar el mínimo, podar.
    """
    partial = numeros[0]
    for i, op in enumerate(operaciones):
        b = numeros[i+1]
        if op == '+':
            partial += b
        elif op == '-':
            partial -= b
        elif op == '*':
            partial *= b
        elif op == '/':
            if b == 0 or partial % b != 0:
                return True
            partial = partial // b

        # Poda si el parcial no puede mejorar el máximo
        if max_val is not None and partial > max_val + 100:
            return True
        # Poda si el parcial no puede mejorar el mínimo
        if min_val is not None and partial < min_val - 100:
            return True
    return False

def resolver_operaciones(operaciones, numeros, operaciones_calculadas):
    nums = list(numeros)
    ops = list(operaciones)
    i = 0
    while i < len(ops):
        if ops[i] in ('*', '/'):
            a, b = nums[i], nums[i+1]
            clave = clave_normalizada(a, ops[i], b)
            if clave in operaciones_calculadas:
                res = operaciones_calculadas[clave]
            else:
                if ops[i] == '*':
                    res = a * b
                elif b == 0 or a % b != 0:
                    return None, operaciones_calculadas
                else:
                    res = a // b
                operaciones_calculadas[clave] = res
            nums[i] = res
            del nums[i+1]
            del ops[i]
        else:
            i += 1

    result = nums[0]
    for i, op in enumerate(ops):
        b = nums[i+1]
        clave = clave_normalizada(result, op, b)
        if clave in operaciones_calculadas:
            result = operaciones_calculadas[clave]
        else:
            if op == '+':
                result += b
            elif op == '-':
                result -= b
            operaciones_calculadas[clave] = result
    return result, operaciones_calculadas

def resolver_algoritmo_branch_and_bound(operaciones=['+', '-', '*', '/'], numeros=list(range(1, 10))):
    inicio = time.time()
    combinacion_operaciones = crear_combinaciones_de_operaciones(operaciones)
    combinacion_numeros = crear_combinaciones_numericas(numeros)

    total_revisadas = 0
    total_calculadas = 0
    min_val = None
    max_val = None
    cache_operaciones = {}

    for numeros in combinacion_numeros:
        for operaciones in combinacion_operaciones:
            total_revisadas += 1
            if se_descarta_por_poda(operaciones, numeros, max_val, min_val):
                continue
            try:
                resultado, cache_operaciones = resolver_operaciones(operaciones, numeros, cache_operaciones)
            except ZeroDivisionError:
                continue
            if resultado is not None and es_entero(resultado):
                valor = int(resultado)
                if min_val is None or valor < min_val:
                    min_val = valor
                if max_val is None or valor > max_val:
                    max_val = valor
                total_calculadas += 1

    print(f"🧮 Total combinaciones revisadas: {total_revisadas}")
    print(f"🧮✂️ Total combinaciones calculadas válidas: {total_calculadas}")
    print(f"📉 Mínimo: {min_val}")
    print(f"📈 Máximo: {max_val}")
    print(f"⏱️ Tiempo total: {round(time.time() - inicio, 2)} segundos")

resolver_algoritmo_branch_and_bound()

🧮 Total combinaciones revisadas: 362880
🧮✂️ Total combinaciones calculadas válidas: 65788
📉 Mínimo: -69
📈 Máximo: 77
⏱️ Tiempo total: 0.18 segundos


In [92]:
operaciones_fuerza_bruta = 362880
operaciones_optimizado = 65788

# porcentaje mejora
print(f"🧮 Operaciones alg. fuerza bruta: {operaciones_fuerza_bruta}")
print(f"🧮✂️ Operaciones alg. optimizado: {operaciones_optimizado}")
print(f"📉 Porcentaje de mejora: {round((operaciones_fuerza_bruta - operaciones_optimizado) / operaciones_fuerza_bruta * 100)}%")

🧮 Operaciones alg. fuerza bruta: 362880
🧮✂️ Operaciones alg. optimizado: 65788
📉 Porcentaje de mejora: 82%


**Conclusiones**

Tras aplicar el algoritmo optimizado con **Branch & Bound y poda de ramas**, se han obtenido **los mismos resultados que con el algoritmo por fuerza bruta**:

📉 Mínimo: **-69**
📈 Máximo: **77**

Sin embargo, el número de operaciones simples realizadas se ha reducido drásticamente:

- Operaciones algoritmo fuerza bruta: 362,880
- Operaciones algoritmo optimizado: **65,788**
- Porcentaje de reducción de operaciones: **82%**

Esto demuestra que el algoritmo optimizado mantiene la exactitud de los resultados mientras mejora significativamente la eficiencia computacional al descartar combinaciones innecesarias, permitiendo resolver el problema de forma exacta con un menor consumo de recursos, lo que resulta especialmente útil en problemas de mayor tamaño.

Por otro lado, el algoritmo con poda es exacto en el cálculo del mínimo y máximo posibles, pero **no asegura la generación de todos los enteros intermedios** del rango obtenido. Para garantizar cobertura total del rango, sería necesario utilizar fuerza bruta sin poda o realizar una comprobación posterior de huecos tras el cálculo.

## **7. Calcula la complejidad del algoritmo.**

La complejidad del algoritmo es la misma, la mejora que se hace es al reducir las operaciones, pero seguiría siendo un algoritmo con complejidad **O(n^5)**, siendo n el tamaño del conjunto de cifras posibles.

## **8. Según el problema (y tenga sentido), diseña un juego de datos de entrada aleatorios.**

Para comprobar la diferencia entre tiempos de ejecución y poder hacer estimaciones se calcularán los datos:
- operaciones=['+', '-', '*', '/'], numeros=list(range(1, 10))
- operaciones=['+', '-', '*', '/'], numeros=list(range(1, 30))
- operaciones=['+', '-', '*', '/'], numeros=list(range(1, 50))



## **9. Aplica el algoritmo al juego de datos generado.**

In [93]:
resolver_fuerza_bruta(operaciones=['+', '-', '*', '/'], numeros=list(range(1, 30)))

ℹ️ Total de combinaciones de operaciones: 24
ℹ️ Total de combinaciones numéricas: 14250600
🧮 Total de combinaciones entre números y operaciones: 342014400
📉 Mínimo: -809
📈 Máximo: 837
⏱️ Tiempo fuerza bruta: 1205.65 segundos


In [94]:
resolver_algoritmo_branch_and_bound(operaciones=['+', '-', '*', '/'], numeros=list(range(1, 30)))

🧮 Total combinaciones revisadas: 342014400
🧮✂️ Total combinaciones calculadas válidas: 27778183
📉 Mínimo: -809
📈 Máximo: 837
⏱️ Tiempo total: 127.9 segundos


In [95]:
resolver_fuerza_bruta(operaciones=['+', '-', '*', '/'], numeros=list(range(1, 50)))

ℹ️ Total de combinaciones de operaciones: 24
ℹ️ Total de combinaciones numéricas: 228826080
🧮 Total de combinaciones entre números y operaciones: 5491825920
📉 Mínimo: -2349
📈 Máximo: 2397
⏱️ Tiempo fuerza bruta: 20533.94 segundos


In [96]:
resolver_algoritmo_branch_and_bound(operaciones=['+', '-', '*', '/'], numeros=list(range(1, 50)))

🧮 Total combinaciones revisadas: 5491825920
🧮✂️ Total combinaciones calculadas válidas: 312056258
📉 Mínimo: -2349
📈 Máximo: 2397
⏱️ Tiempo total: 1879.44 segundos


In [112]:
# Prueba de 10 numeros
numero_combinaciones_10_fuerza_bruta = 362880
numero_combinaciones_10_optimizado = 65788
print(f"📉 Porcentaje de mejora 10 números: {round((numero_combinaciones_10_fuerza_bruta - numero_combinaciones_10_optimizado) / numero_combinaciones_10_fuerza_bruta * 100)}%")

📉 Porcentaje de mejora 10 números: 82%


In [113]:
# Prueba de 30 numeros
numero_combinaciones_30_fuerza_bruta = 342014400
numero_combinaciones_30_optimizado = 27778183
print(f"📉 Porcentaje de mejora 30 números: {round((numero_combinaciones_30_fuerza_bruta - numero_combinaciones_30_optimizado) / numero_combinaciones_30_fuerza_bruta * 100)}%")

📉 Porcentaje de mejora 30 números: 92%


In [ ]:
# Prueba de 50 numeros
numero_combinaciones_50_fuerza_bruta = 5491825920
numero_combinaciones_50_optimizado = 312056258
print(f"📉 Porcentaje de mejora 50 números: {round((numero_combinaciones_50_fuerza_bruta - numero_combinaciones_50_optimizado) / numero_combinaciones_50_fuerza_bruta * 100)}%")

📉 Porcentaje de mejora 50 números: 94%


## Conclusiones resultados

Al realizar pruebas con rangos de 10, 30 y 50 números de entrada, se observa que la aplicación de poda y restricciones permite reducir drásticamente el número de combinaciones evaluadas sin alterar el resultado final de mínimo y máximo encontrados. 

Los resultados muestran:
- Con 10 números, se obtiene un 82% de reducción de combinaciones calculadas respecto a fuerza bruta.
- Con 30 números, se alcanza un 92% de reducción.
- Con 50 números, la reducción sube hasta un 94%.

Este patrón demuestra que cuanto mayor es el rango de números de entrada, mayor es la eficiencia de la poda aplicada en el algoritmo, ya que el espacio combinatorio crece exponencialmente y la poda permite descartar rápidamente combinaciones inviables o redundantes antes de su cálculo.

De esta forma, el algoritmo se vuelve escalable y aplicable a rangos más amplios, manteniendo la calidad de los resultados finales (mínimo y máximo obtenidos) y reduciendo drásticamente el número de operaciones simples necesarias, cumpliendo con los objetivos de optimización exigidos en la asignatura.

## **10. Describe brevemente las lineas de como crees que es posible avanzar en el estudio del problema. Ten en cuenta incluso posibles variaciones del problema y/o variaciones al alza del tamaño.**

Una vez optimizado el algoritmo inicial, existen varias líneas para seguir profundizando en el estudio y mejora del problema:

- **Explorar otros algoritmos:** Probar algoritmos de optimización como algoritmos genéticos, programación dinámica y búsqueda local, que permiten resolver instancias de mayor tamaño de forma eficiente al incrementar el rango de números o el número de operaciones aunque se pierda la precisión.

- **Paralelización del algoritmo:** Implementar una versión concurrente o distribuida permitiría aprovechar múltiples núcleos o incluso varias máquinas para reducir aún más el tiempo de ejecución, especialmente útil cuando el tamaño del conjunto de números crece (por ejemplo, con más de 100 elementos).

- **Uso de memoria local**: uno de los grandes problemas que he tenido ha sido que cuando utilizaba funciones que devolvian un array de elementos provocaba un consumo excesivo de la memoria RAM del sistema, por lo que se podría optimizar para particionar los datos y jugar con la carga entre ROM y RAM.

- **Escalado para conjuntos más grandes:** A medida que se aumenta la cantidad de números o de operaciones, el número de combinaciones crece de forma exponencial. Por tanto, sería interesante diseñar una estrategias de muestreo que reduzcan el espacio de búsqueda manteniendo una buena precisión.

- **Representación más eficiente de datos:** Cambiar las estructuras de datos actuales por implementaciones más ligeras (como tuplas comprimidas, generadores o arrays con numpy) podría reducir significativamente el uso de memoria, lo cual es clave al escalar a millones de combinaciones.